ramips: sunvalley - allow larger sysupgrade

arrmo opened the pull request:
Calculated total free space, set to match
Checked on hardware, no issues seen.

Signed-off-by: Russell Morris <rmorris@rkmorris.us>

Commit:
ramips: sunvalley - allow larger sysupgrade
Aligned to size of mtd-concat partition (firmware)
  - in this device we have mtd-concat driver that joins multiple flash partitions
  - since sysupgrade works with mtd devices the rootfs partition is already joined
  - we can use a bigger sysupgrade image than factory/TFTP install images
Checked on hardware, no issues seen.
No modifications to images other than sysupgrade (i.e. TFTP / recovery images not touched).

Signed-off-by: Russell Morris <rmorris@rkmorris.us>

xabolcs commented:
NAK from me!
Initial install through TFTP will render (IMHO) the device to a paperwheight, because of overflowing to the beginning of the flash, where the bootloader resides.

You already know that feeling! 🙃

If you want to do this space extension, then you should create a recipe which cuts the `rootfs` at 6M and concatenates the remainder to `kernel` like it would be written by `mtd-split`.

xabolcs commented:
Reading through the device page of [RP-WD03](https://openwrt.org/toh/ravpower/rp-wd03#flash_layout), the TFTP recovery method works up to 6144k + 1536k = **7680k** while maintaining compatibility with OEM flash layout: `params` / `u-boot-env` should be kept intact (and of course `loader` need it's own partition - 64k).

Flashing `rootfs` bigger than 1536k through TFTP, it will overwrite the `params` partition. :roll_eyes: 

The mentioned recipe should do the following:
* should split `$$(sysupgrade_bin)` at 6144k
* the first part should make it's way as `rootfs`, and shouldn't exceed 6144k
* the second part should be appended to `append-loader-okli $(1)`, do some some `mtd-concat` compatibility hacks, and we got the flashable `kernel`. This shouldn't exceed 1536k!

arrmo commented:
Hmmm ... you are very likely right, I need to dig into this more. My issue is specifically sysupgrade - and it can be larger, no? So we need different size restrictions for TFTP vs. sysupgrade images?

Thanks!

xabolcs commented:
sysupgrade = kernel + rootfs.
TFTP recovery = OKLI Loader (64k) + kernel + rootfs

TFTP works up to 7680k which is only 2x64k less than the maximal 7744k.

If you allow sysupgrade up to 7744k, then those images won't work for TFTP recovery - you had to different packages list and/or kernel config for TFTP and for sysupgrade. 🤷

Do you have size problems? Why don't you use [extroot](https://openwrt.org/docs/guide-user/additional-software/extroot_configuration)?

arrmo commented:
> If you allow sysupgrade up to 7744k, then those images won't work for TFTP recovery - you had to different packages list and/or kernel config for TFTP and for sysupgrade. 🤷

That makes sense! So set the maximum to 7680k, as you note?

> Do you have size problems? Why don't you use [extroot](https://openwrt.org/docs/guide-user/additional-software/extroot_configuration)?

I do (have a size issue), unfortunately. Updated to the lastest code (master branch), and well over the current value that is in there (i.e. 6144k). 7680k would be fine, and avoid going through extroot "fun" 😉. You good with 7680k (vs. 7744k)?

Thanks!

xabolcs commented:
> So set the maximum to 7680k, as you note?
> 

The current recipe is safe up to 6144k, the current value.
The recipe I mentioned is safe up to 7680k.
7744k is impossible for TFTP recovery with maintaining OEM partitions compatibility.

Without implementing the new recipe, you shouldn't set the maximum above 6144k.



> I do (have a size issue), unfortunately. Updated to the lastest code (master branch), and well over the current value that is in there (i.e. 6144k).

Thats sad.

But I simply just don't understand what you wrote.
`rootfs.bin` (kernel + rootfs) sizes are:
* current master has 4864k `hootoo_ht-tm05-squashfs-rootfs.bin`, 
* 21.02.0-rc3 has 5120k for the same `rootfs.bin`

Do you use ImageBuilder? :thinking:

arrmo commented:
> Do you use ImageBuilder? 🤔

Nope, straight up build (make). It seemed to grow quite a bit in the latest master => now rootfs.bin is 7169k. I did add in one thing (bind-dig) ... not sure if that added a whole bunch of dependencies or not. Rebuilding, with bind-dig removed ... nope, still rootfs.bin is 7169k. And actually, initramfs-kernel.bin is 6928k.

arrmo commented:
Looking through the code, it seems that IMAGE_SIZE is the same for TFTP or sysupgrade ... but now that we have sections of flash "stitched" together, it's almost like we need 2 different values. Agreed? But that's a lot bigger code changes 😞.

Thanks!

xabolcs commented:
Just to note: you don't have to stick to `IMAGE_SIZE`. You could use arbitrary numbers after `check-size` in the recipe.

So I don't see why are you interested in 2 different values, instead thinking about a method which safely creates a `kernel.bin` and `rootfs.bin` from an "oversized" (6144 k <= x <= 7680k) `$$(sysupgrade_bin)`.

arrmo commented:
> Just to note: you don't have to stick to `IMAGE_SIZE`. You could use arbitrary numbers after `check-size` in the recipe.

I like that, good idea!



> So I don't see why are you interested in 2 different values, instead thinking about a method which safely creates a `kernel.bin` and `rootfs.bin` from an "oversized" (6144 k <= x <= 7680k) `$$(sysupgrade_bin)`.

Sounds good ... I think 😜. You lost me there, not quite sure what you're after. Can you clarify a bit? Thanks!

xabolcs commented:
>  You lost me there, not quite sure what you're after. Can you clarify a bit? Thanks!

I already wrote it earlier, but let my retry!

The maximum size the TFTP recovery is able to flash is 1536k (`kernel`) + 6144k (`rootfs`) = 7680k

Currently the OKLI loader goes into `kernel`, and the real kernel + rootfs (`$$sysupgrade_bin`, as [`sysupgrade_bin := append-kernel | append-rootfs | pad-rootfs`](https://github.com/openwrt/openwrt/blob/5c88bf896ec1be9129292aeb522bfc0f75ec9898/target/linux/ramips/image/Makefile#L168)) goes into `kernel`.
And to prevent the "I don't always overflow on TFTP recovery images ..." meme, there is a size check for 6144k for the whole.

So, if you want to install an image which is bigger than 6144k (the OEM `​Rootfs` partition), then you have to split that image into `kernel` + `rootfs`.

Current 21.02 sizes are
* `hootoo_ht-tm05-squashfs-kernel.bin`: 3.3k (OKLI Loader)
* `hootoo_ht-tm05-squashfs-rootfs.bin`: 5120k (kernel + rootfs + padding)


Visualized space usage:
```
Legend:
O: OKLI Loader + `pad-to 64k`
K: OpenWrt kernel
R: OpenWrt rootfs

rootfs: used 4k (64k) / max 1536k
---------------
O              |
---------------

kernel: used 5120k / max 6144k
--------------------------------------------------------------
KKKKKKKKKKKKKKKKKKKKRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRR          |
--------------------------------------------------------------
```

And when you put all your bloat into your image, then you pass 6144k, you had to split that image into those two TFTP recovery file: `kernel` and `rootfs`.

```
Legend:
O: OKLI Loader + `pad-to 64k`
H: first part (up to 6144k) of sysupgrade_bin (OpenWrt kernel + OpenWrt rootfs)
T: second part (the remainder) of sysupgrade_bin (OpenWrt kernel + OpenWrt rootfs)

rootfs: used xxxxk / max 1536k
---------------
OTTTTTTTTTTT  |
---------------

kernel: used 6144k / max 6144k
--------------------------------------------------------------
HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH|
--------------------------------------------------------------
```

Now you understand what I wrote in my previous comments?

arrmo commented:
> Now you understand what I wrote in my previous comments?

Yes, thanks for the details - appreciated! I was slow to reply, as I was trying to step through, see why the image size was blowing up so much. So I started from scratch, added my packages ... here is what I found,
![image](https://user-images.githubusercontent.com/8756384/129655992-f8a9b991-7d38-47eb-8a5d-78b677fcb261.png)

I have no idea what kmod-crypto-kpp is, but it's absolutely blowing up the image size. I didn't select it, but it seems to be around. What's odd - now I can't seem to get rid of it. If I add all the items before it in the list, it's there automatically! But, if I add them one by one ... nope. What?!?!

I even tried a make clean - no joy. Make dirclean - still not. Can't seem to get rid of it now. Thinking it's WireGuard related, but it's not there stepping through one by one?

Pulling my hair out! 😆

arrmo commented:
Let me open a side discussion to debug the image size - seeing some real odd items there. Thinking something isn't right, or at least consistent => but it's what started this PR / debugging. We can work it out there, then come back here 😄.

Thanks!

arrmo commented:
OK, here is my post about [image size](https://forum.openwrt.org/t/ramips-image-size-exploding/103972?u=arrmo). Will also create one for debugging the Makefile (but let me fiddle with that a bit first 😆).

Thanks!

arrmo commented:
And the second post, about this [Device ](https://forum.openwrt.org/t/ramips-sunvalley-permit-larger-sysupgrade/103976)(recipe). Thanks!

adschm commented:
Commit title needs to be updated.

arrmo commented:
> Commit title needs to be updated.

Agreed! You OK with me updating it similar to the discussion that I think we just closed on (i.e. similar to "Ramips sunvalley, Permit Larger sysupgrade")?

Thanks!

arrmo commented:
Updated the naming (title above, and commit), reverted IMAGE_SIZE, only allow sysupgrade to be larger (and matched to firmware size, for mtd-concat partition).

Thanks!

bobafetthotmail commented:
I think you need to explain better in the commit message why you are doing this. There is no size limit in commit description and it is usually better to explain well why you do the change

Like:
"in this device we have mtd-concat driver that joins multiple flash partitions and since sysupgrade works with mtd devices the rootfs partition is already joined and we can use a bigger sysupgrade image than factory/TFTP install images"

arrmo commented:
Updated - and thanks @bobafetthotmail! I blatently stole some of your text above 😆.

Yell if you see any other issues. Thanks again.

adschm commented:
Commit title still not following formal requirements.

arrmo commented:
OK, let me do some more digging, sorry!

I checked [here](https://openwrt.org/submitting-patches), but I may be missing what I need to change in the title. Thanks!

xabolcs commented:
Read it more carefully: @arrmo , you updated the PR's title ... but there is another title around! The **commit**'s title! :wink:

arrmo commented:
Dang it ... LOL. Thanks! I missed the first word in the commit, my bad. I tried to make them them same, but tripped over myself 😜.

Appreciate it!

neheb commented:
ping @Ansuel this looks ready
